crates/configuration/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+                .expect("Failed to read snfoundry.toml config file")
+                .parse::<Value>()
+                .expect("Failed to parse snfoundry.toml config file");
+
+            let profile = get_profile(&raw_config, T::tool_name(), profile)?;
+            T::from_raw(&profile)
+        }
+        None => Ok(T::default()),
+    }
+}
+
+pub fn search_config_upwards_relative_to(current_dir: &Utf8PathBuf) -> Result<Utf8PathBuf> {
+    current_dir
+        .ancestors()
+        .find(|path| fs::metadata(path.join(CONFIG_FILENAME)).is_ok())
+        .map(|path| path.join(CONFIG_FILENAME))
+        .ok_or_else(|| {
+            anyhow!(
+                "Failed to find snfoundry.toml - not found in current nor any parent directories"
+            )
+        })
+}
+
+pub fn find_config_file() -> Result<Utf8PathBuf> {
+    search_config_upwards_relative_to(&Utf8PathBuf::try_from(
+        std::env::current_dir().expect("Failed to get current directory"),
+    )?)
+}
+
+pub trait PropertyFromCastConfig: Sized {
+    fn from_toml_value(value: &Value) -> Option<Self>;
+}
+
+impl PropertyFromCastConfig for String {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_str().map(std::borrow::ToOwned::to_owned)
+    }
+}
+
+impl PropertyFromCastConfig for Utf8PathBuf {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_str().map(Utf8PathBuf::from)
+    }
+}
+
+impl PropertyFromCastConfig for u8 {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_integer().and_then(|i| i.try_into().ok())
+    }
+}
+
+impl PropertyFromCastConfig for u16 {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_integer().and_then(|i| i.try_into().ok())
+    }
+}
+
+impl<T> PropertyFromCastConfig for Option<T>
+where
+    T: PropertyFromCastConfig,
+{
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        T::from_toml_value(value).map(Some)
+    }
+}
+
+pub fn get_property<T>(entries: &Value, field: &str) -> Option<T>
+where
+    T: PropertyFromCastConfig + Default,
+{
+    entries.get(field).and_then(T::from_toml_value)
+}
+
+#[must_use]
+pub fn copy_config_to_tempdir(src_path: &str, additional_path: Option<&str>) -> TempDir {
+    let temp_dir = tempdir().expect("Failed to create a temporary directory");
+    if let Some(dir) = additional_path {
+        let path = temp_dir.path().join(dir);
+        fs::create_dir_all(path).expect("Failed to create directories in temp dir");
+    };
+    let temp_dir_file_path = temp_dir.path().join(CONFIG_FILENAME);
+    fs::copy(src_path, temp_dir_file_path).expect("Failed to copy config file to temp dir");
+    temp_dir
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use tempfile::tempdir;
+
+    use super::*;
+
+    #[test]
+    fn find_config_in_current_dir() {
+        let tempdir =
+            copy_config_to_tempdir("../sncast/tests/data/files/stubtool_snfoundry.toml", None);
```

Comment:
I guess it would make sense to hold test files for this crate in this crate

Reply:
Not sure what do you mean 🤔

Reply:
we are in a different crate (configuration), so should we use a file (stubtool_snfoundry.toml) that sits in a different crate? I think the file could be moved

Reply:
Ah yes sure, I intended to do so, forgot about it

Reply:
Done